### PR TITLE
Feat: Basic Docker Container to run on Bacalhau

### LINKED ISSUE #1: Feat: create a basic docker container which runs on Bacalhau
# User Story
As a developer I want to have a basic framework ready on how to run code on bacalhau so that I can speed up my development process.

# Status Quo
Currently no job or code can be deployed on Bacalhau

# Acceptance Criteria

- [x] A simple docker container that read data from IPFS and logs it.
- [x] Script to easily publish docker container to registry
- [x] Script to easily push run Bacalhau Jobs
- [ ] Considers future Javascript SES implementation

### PR DESCRIPTION
**Pull request by codebuster22**
# What I have done
1. Created a basic docker container which runs on bacalhau
2. Reads input from `./inputs` directory.
3. Takes a string and convert it to dynamic function from input.
4. Stores the output of function in `./outputs` under `outputs.json` file.
5. Deployed the docker image on docker hub.

closes #1

### CODE REVIEW

File: packages/off-chain/README.md [hunk 1]
```diff
@@ -1,3 +1,55 @@
 # off-chain
 
-run example script "doSomething.ts with `node-ts doSomething.ts`
+## Links
+Docker Image: mihirsinhparmar/ie-bac
+
+Run jon on bacalhau
+```
+bacalhau docker run mihirsinhparmar/ie-bac
```
Comment (fabianschu) [inline packages/off-chain/README.md]: This command doesn't work for me "Error while executing job".

Is that the same amd64 issue?

This is what I get if I use `bacalhau describe ...`:

```
APIVersion: V1beta1
Metadata:
  ClientID: f4d3ca22f82233396e43ff64383a6f8a80e2a53915d8ff05660292bdd2e111bf
  CreatedAt: "2023-01-02T22:05:06.834907625Z"
  ID: e7faeac5-199d-470f-a417-2bbc2f0cfb06
Spec:
  Deal:
    Concurrency: 1
  Docker:
    Image: mihirsinhparmar/ie-bac
  Engine: Docker
  ExecutionPlan:
    ShardsTotal: 1
  Language:
    JobContext: {}
  Publisher: Estuary
  Resources:
    GPU: ""
  Sharding:
    BatchSize: 1
    GlobPatternBasePath: /inputs
  Timeout: 1800
  Verifier: Noop
  Wasm: {}
  outputs:
  - Name: outputs
    StorageSource: IPFS
    path: /outputs
Status:
  JobState:
    Nodes:
      QmVAb7r2pKWCuyLpYWoZr9syhhFnTWeFaByHdb8PkkhLQG:
        Shards:
          "0":
            NodeId: QmVAb7r2pKWCuyLpYWoZr9syhhFnTWeFaByHdb8PkkhLQG
            PublishedResults: {}
            State: Cancelled
            VerificationResult: {}
      QmXaXu9N5GNetatsvwnTfQqNtSeKAD6uCmarbh3LMRYAcF:
        Shards:
          "0":
            NodeId: QmXaXu9N5GNetatsvwnTfQqNtSeKAD6uCmarbh3LMRYAcF
            PublishedResults: {}
            State: Cancelled
            VerificationResult: {}
      QmYgxZiySj3MRkwLSL4X2MF5F9f2PMhAE3LV49XkfNL1o3:
        Shards:
          "0":
            NodeId: QmYgxZiySj3MRkwLSL4X2MF5F9f2PMhAE3LV49XkfNL1o3
            PublishedResults: {}
            State: Error
            Status: Received ActionComputeError from node QmYgxZiySj3MRkwLSL4X2MF5F9f2PMhAE3LV49XkfNL1o3
              that has not bid on this shard
            VerificationResult: {}
      QmdZQ7ZbhnvWY1J12XYKGHApJ6aufKyLNSvf8jZBrBaAVL:
        Shards:
          "0":
            NodeId: QmdZQ7ZbhnvWY1J12XYKGHApJ6aufKyLNSvf8jZBrBaAVL
            PublishedResults: {}
            State: Cancelled
            VerificationResult: {}
  Requester:
    RequesterNodeID: QmdZQ7ZbhnvWY1J12XYKGHApJ6aufKyLNSvf8jZBrBaAVL
    RequesterPublicKey: CAASpgIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDVRKPgCfY2fgfrkHkFjeWcqno+MDpmp8DgVaY672BqJl/dZFNU9lBg2P8Znh8OTtHPPBUBk566vU3KchjW7m3uK4OudXrYEfSfEPnCGmL6GuLiZjLf+eXGEez7qPaoYqo06gD8ROdD8VVse27E96LlrpD1xKshHhqQTxKoq1y6Rx4DpbkSt966BumovWJ70w+Nt9ZkPPydRCxVnyWS1khECFQxp5Ep3NbbKtxHNX5HeULzXN5q0EQO39UN6iBhiI34eZkH7PoAm3Vk5xns//FjTAvQw6wZUu8LwvZTaihs+upx2zZysq6CEBKoeNZqed9+Tf+qHow0P5pxmiu+or+DAgMBAAE=
```

File: packages/off-chain/inputs/dummyInput.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "data": {
+    "param1": 10,
+    "param2": 20
+  },
+  "function": {
+      "params": ["param1", "param2"],
+      "code": "return param1 + param2"
+  }
+}
```
Comment (fabianschu) [inline packages/off-chain/inputs/dummyInput.json:10]: I get that this is just for testing purposes, but just to make sure: later on the actual contribution data and the definition of the IE function should be two separate inputs.

File: packages/off-chain/scripts/main.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  // checking if the input contain function property or not
+  // custom impact evaluator function should be stored with property 'function'
+  if (inputs.function !== undefined) {
+    // get the function logic code
+    const functionCode = inputs.function.code;
+    // get the array of params
+    const params = inputs.function.params;
+    // instantiate function
+    ieFunction = new Function(...params, functionCode);
```
Comment (fabianschu) [inline packages/off-chain/scripts/main.ts]: nice!

File: packages/off-chain/scripts/main.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      result: result,
+      inputs: inputs.length === 0 ? "No Input found" : inputs,
+    };
+    const dataToBeWritten = JSON.stringify(output, undefined, 4);
+    fs.writeFileSync(outputPath, dataToBeWritten);
+  } catch (e) {
+    // some issue faced while writing
+    console.log("Writing output failed");
+    console.log(e);
+  }
```
Comment (fabianschu) [inline packages/off-chain/scripts/main.ts]: A heads up: the exact format of the pinned output file will likely have to be adjusted to work with the astrodrop frontend. I'm referring to [this](https://github.com/Astrodrop/shrine/blob/main/src/Shrine.sol#L143) parameter, which might need to point to a ipfs file that follows some specific format. Specifics tbd. Btw do you know if it is possible to pin more than one files to ipfs from bacalhau?

File: packages/off-chain/scripts/main.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
+// to be run on bacalhau via Docker container
+// file system is used to fetch the data from ../inputs/ directory
+import * as fs from "fs";
+import * as path from "path";
+import {successOutput, failedOutput} from "./outputSchema.js";
```
Comment (fabianschu) [inline packages/off-chain/scripts/main.ts]: These imports are not used as far as I understand. Are they necessary?

File: packages/off-chain/README.md (hunk 1 above)
Comment (codebuster22) [reply packages/off-chain/README.md]: yeah, this is something that I have been facing multiple times. That is, it is not deterministic.

It's not an issue of `amd64`.

When I run it, it runs.

Here's the log:
```
0xmihir@Mihirs-MacBook-Pro off-chain % bacalhau docker run mihirsinhparmar/ie-bac   
Job successfully submitted. Job ID: ace03c38-1ed8-411b-9ffd-276dacf9c256
Checking job status... (Enter Ctrl+C to exit at any time, your job will continue running):

               Creating job for submission ... done ✅
               Finding node(s) for the job ... done ✅
                     Node accepted the job ... done ✅
           Job finished, verifying results ... done ✅
              Results accepted, publishing ... Job Results By Node:

To download the results, execute:
  bacalhau get ace03c38-1ed8-411b-9ffd-276dacf9c256

To get more details about the run, execute:
  bacalhau describe ace03c38-1ed8-411b-9ffd-276dacf9c256
```

File: packages/off-chain/inputs/dummyInput.json (hunk 2 above)
Comment (codebuster22) [reply packages/off-chain/inputs/dummyInput.json:10]: Yes, the parameters and the function definition should be in two separate properties.

The reason being, `eval` runs the code directly, so we won't be able to pass any parameters. Also, using `eval` is a security risk as any code will run in the global context and can be dangerous (though, no high risk processing is done here).

Whereas `Function` class, accepts parameters and creates an instance of function, which we can invoke anytime.

Resources:
[Function()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/Function#:~:text=Description-,Function()%20constructor,performance%20issues%20as%20eval()%20.)
[eval](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval)
[eval vs Function()](https://www.educative.io/answers/eval-vs-function-in-javascript)

File: packages/off-chain/scripts/main.ts (hunk 3 above)
Comment (codebuster22) [reply packages/off-chain/scripts/main.ts]: 🤓

File: packages/off-chain/scripts/main.ts (hunk 4 above)
Comment (codebuster22) [reply packages/off-chain/scripts/main.ts]: > Btw do you know if it is possible to pin more than one files to ipfs from bacalhau?

Yes, we can pin multiple files to IPFS via bacalhau.

> the exact format of the pinned output file will likely have to be adjusted to work with the astrodrop frontend

Yes, I have created a temporary schema of output. Once we have more clarity on how both modules will communicate, we will define the schema in `./outputSchema.ts` file. And all the output (Success or Error) will be stored in that defined schema. To ensure, both modules can understand each other.

File: packages/off-chain/scripts/main.ts (hunk 5 above)
Comment (codebuster22) [reply packages/off-chain/scripts/main.ts]: Yes, currently it's not, but this imports are a defined schema or structure in which output should be stored.

File: packages/off-chain/README.md (hunk 1 above)
Comment (codebuster22) [reply packages/off-chain/README.md]: now the weird part is, I run the command
`bacalhau docker run mihirsinhparmar/ie-bac`

the first time, it fails, the second time it runs
```
FIRST TRY
0xmihir@Mihirs-MacBook-Pro off-chain % bacalhau docker run mihirsinhparmar/ie-bac
Job successfully submitted. Job ID: d7d7abc2-03e2-47ec-b249-14a5f4a0c083
Checking job status... (Enter Ctrl+C to exit at any time, your job will continue running):

               Creating job for submission ... done ✅
               Finding node(s) for the job ... done ✅
                     Node accepted the job ... done ✅
           Error while executing the job.

Job Results By Node:

To download the results, execute:
  bacalhau get d7d7abc2-03e2-47ec-b249-14a5f4a0c083

To get more details about the run, execute:
  bacalhau describe d7d7abc2-03e2-47ec-b249-14a5f4a0c083

SECOND TRY (ran 3 seconds after first try)
0xmihir@Mihirs-MacBook-Pro off-chain % bacalhau docker run mihirsinhparmar/ie-bac
Job successfully submitted. Job ID: b5ce2ca0-7f2f-4c78-8abe-2ac1815bdb87
Checking job status... (Enter Ctrl+C to exit at any time, your job will continue running):

               Creating job for submission ... done ✅
               Finding node(s) for the job ... done ✅
                     Node accepted the job ... done ✅
           Job finished, verifying results ... done ✅
              Results accepted, publishing ... Job Results By Node:

To download the results, execute:
  bacalhau get b5ce2ca0-7f2f-4c78-8abe-2ac1815bdb87

To get more details about the run, execute:
  bacalhau describe b5ce2ca0-7f2f-4c78-8abe-2ac1815bdb87
```

Comment (fabianschu) [reply packages/off-chain/README.md]: Weird, I don't get it to successfully execute the job 🤔 

I rebuilt the docker container locally (`docker build ..`; `docker tag ..`), pushed it to `fabianschu/ex1` and now I can execute it succesfully via `bacalhau docker run fabianschu/ex1`. I'm a bit puzzled tbh. Isn't that exactly the same problem that you were facing initially? Feels like an issue on the bacalhau side of things. 

I don't think it should stop you from moving on with the implementation but I'm wondering if we should ask about it in the grant slack channel. Wdyt?

Comment (fabianschu) [reply packages/off-chain/README.md]: Weird, I don't get it to successfully execute the job 🤔

I rebuilt the docker container locally with the shell script and now I can execute it succesfully via bacalhau docker run fabianschu/ex3. I'm a bit puzzled tbh. Isn't that exactly the same problem that you were facing initially? Feels like an issue on the bacalhau side of things.

I don't think it should stop you from moving on with the implementation but I'm wondering if we should ask about it in the grant slack channel. Wdyt?

**Review by fabianschu — APPROVED** [review]
nice! 

I just added one comment about the bacalhau error, but imo this can be merged.

File: packages/off-chain/README.md (hunk 1 above)
Comment (codebuster22) [reply packages/off-chain/README.md]: I have raised the question on slack channel.

**Pull request merged**